File: VotingSystem.py
"""
✨ USA Voting System 2024 ✨
Just name/surname verification, nothing fancy
"""

def print_separator():
    print("\n" + "=" * 50 + "\n")

def fancy_print(message):
    print(f"  {message}")

def register_candidates():
    candidates = [
        "1. Donald Trump     ",  
        "2. Kamala Harris    ",
        "3. Cornel West      ",
        "4. Jill Stein       ",
        "5. Chase Oliver     ",
        "6. Claudia De la Cruz"
    ]
    
    fancy_print("Current Candidates")
    print_separator()
    for candidate in candidates:
        print(f"  {candidate}")
    print_separator()
    return candidates

def cast_vote(candidate_list):
    voted_people = {}  
    vote_counts = {candidate: 0 for candidate in candidate_list}
    
    fancy_print("Let's start voting!")
    
    while True:
        print("\n" + "-" * 30)
        name = input("Enter your name (or 'exit' to finish): ").strip().title()
        
        if name.lower() == 'exit':
            break
            
        surname = input("Enter your surname: ").strip().title()
        voter_key = f"{name} {surname}"
            
        if voter_key in voted_people:
            print("⚠  Hey, looks like you already voted!")
            continue
            
        print("\nPick your candidate:")
        for candidate in candidate_list:
            print(f"  {candidate}")
            
        vote = input("\n📝 Which number? ")
        valid_numbers = [c.split('.')[0] for c in candidate_list]
        
        if vote in valid_numbers:
            for candidate in candidate_list:
                if candidate.startswith(vote):
                    vote_counts[candidate] += 1
                    voted_people[voter_key] = True
                    print(f"Thanks {name}! Vote registered for {candidate}")
                    break
        else:
            print("Oops! That's not a valid candidate number!")
    
    return vote_counts

def display_results(votes):
    print_separator()
    fancy_print("Here's how everyone did:")
    print_separator()
    
    total_votes = sum(votes.values())
    if total_votes == 0:
        total_votes = 1
    
    for candidate, num_votes in votes.items():
        percentage = (num_votes / total_votes) * 100
        bar = "█" * int(percentage // 5) + "▒" * (20 - int(percentage // 5))
        print(f"{candidate}")
        print(f"{bar} {num_votes} votes ({percentage:.1f}%)\n")
    
    return votes

def run_second_round(tied_candidates):
    fancy_print("Time for a tie-breaker!")
    print("\nOnly these candidates are left:")
    
    voted_people = {}
    
    while True:
        print_separator()
        for candidate in tied_candidates:
            print(f"  {candidate}")
        
        round2_votes = {candidate: 0 for candidate in tied_candidates}

        while True:
            print("\n" + "-" * 30)
            name = input("Enter your name (or 'exit' to finish): ").strip().title()
            if name.lower() == 'exit':
                break
                
            surname = input("Enter your surname: ").strip().title()
            voter_key = f"{name} {surname}"
            
            if voter_key in voted_people:
                print("  You already voted in this round!")
                continue
                
            vote = input("📝 Candidate number: ")
            if vote in [c.split('.')[0] for c in tied_candidates]:
                for candidate in tied_candidates:
                    if candidate.startswith(vote):
                        round2_votes[candidate] += 1
                        voted_people[voter_key] = True
                        print(f"Got it! Vote cast for {candidate}")
                        break
            else:
                print(" That's not a valid choice!")

        display_results(round2_votes)
        
        max_votes = max(round2_votes.values())
        winners = [c for c, v in round2_votes.items() if v == max_votes]

        if len(winners) == 1:
            print("\n We have a winner!")
            fancy_print(f"Congratulations {winners[0]}!")
            break
        else:
            print("\n Still tied! Let's go again...")

def main():
    print("\n" + "=" * 50)
    fancy_print("Welcome to USA Voting System 2024")
    print("=" * 50 + "\n")
    
    candidates = register_candidates()
    votes = cast_vote(candidates)
    final_votes = display_results(votes)

    max_votes = max(final_votes.values())
    winners = [c for c, v in final_votes.items() if v == max_votes]
    
    if len(winners) > 1:
        print("\n Looks like we have a tie!")
        run_second_round(winners)
    else:
        print("\n We have a winner!")
        fancy_print(f"Congratulations {winners[0]}!")

if __name__ == "__main__":
    main()

File: VotingSystem.js
const readline = require('readline').createInterface({
    input: process.stdin,
    output: process.stdout
});

function askQuestion(query) {
    return new Promise(resolve => readline.question(query, resolve));
}

function printSeparator() {
    console.log("\n" + "=".repeat(50) + "\n");
}

function fancyPrint(message) {
    console.log("  " + message);
}

function registerCandidates() {
    const candidates = [
        "1. Donald Trump     ",
        "2. Kamala Harris    ",
        "3. Cornel West      ",
        "4. Jill Stein       ",
        "5. Chase Oliver     ",
        "6. Claudia De la Cruz"
    ];

    fancyPrint("Current Candidates");
    printSeparator();
    candidates.forEach(candidate => console.log("  " + candidate));
    printSeparator();
    return candidates;
}

async function castVote(candidateList) {
    const votedPeople = {};
    const voteCounts = Object.fromEntries(candidateList.map(candidate => [candidate, 0]));

    fancyPrint("Let's start voting!");

    while (true) {
        console.log("\n" + "-".repeat(30));
        const name = (await askQuestion("Enter your name (or 'exit' to finish): ")).trim();
        
        if (name.toLowerCase() === 'exit') break;
        
        const surname = (await askQuestion("Enter your surname: ")).trim();
        const voterKey = `${name} ${surname}`;
        
        if (votedPeople[voterKey]) {
            console.log("⚠  Hey, looks like you already voted!");
            continue;
        }
        
        console.log("\nPick your candidate:");
        candidateList.forEach(candidate => console.log("  " + candidate));
        
        const vote = (await askQuestion("\n📝 Which number? ")).trim();
        const validNumbers = candidateList.map(candidate => candidate.split('.')[0]);
        
        if (validNumbers.includes(vote)) {
            for (let candidate of candidateList) {
                if (candidate.startsWith(vote)) {
                    voteCounts[candidate]++;
                    votedPeople[voterKey] = true;
                    console.log(`Thanks ${name}! Vote registered for ${candidate}`);
                    break;
                }
            }
        } else {
            console.log("Oops! That's not a valid candidate number!");
        }
    }
    return voteCounts;
}

function displayResults(votes) {
    printSeparator();
    fancyPrint("Here's how everyone did:");
    printSeparator();

    const totalVotes = Object.values(votes).reduce((a, b) => a + b, 0) || 1;

    for (const [candidate, numVotes] of Object.entries(votes)) {
        const percentage = (numVotes / totalVotes) * 100;
        const bar = "█".repeat(Math.floor(percentage / 5)) + "▒".repeat(20 - Math.floor(percentage / 5));
        console.log(`${candidate}`);
        console.log(`${bar} ${numVotes} votes (${percentage.toFixed(1)}%)\n`);
    }

    return votes;
}

async function runSecondRound(tiedCandidates) {
    fancyPrint("Time for a tie-breaker!");
    console.log("\nOnly these candidates are left:");

    const votedPeople = {};

    while (true) {
        printSeparator();
        tiedCandidates.forEach(candidate => console.log("  " + candidate));
        
        const round2Votes = Object.fromEntries(tiedCandidates.map(candidate => [candidate, 0]));

        while (true) {
            console.log("\n" + "-".repeat(30));
            const name = (await askQuestion("Enter your name (or 'exit' to finish): ")).trim();
            if (name.toLowerCase() === 'exit') break;

            const surname = (await askQuestion("Enter your surname: ")).trim();
            const voterKey = `${name} ${surname}`;
            
            if (votedPeople[voterKey]) {
                console.log("  You already voted in this round!");
                continue;
            }
            
            const vote = (await askQuestion("📝 Candidate number: ")).trim();
            const validNumbers = tiedCandidates.map(candidate => candidate.split('.')[0]);
            
            if (validNumbers.includes(vote)) {
                for (let candidate of tiedCandidates) {
                    if (candidate.startsWith(vote)) {
                        round2Votes[candidate]++;
                        votedPeople[voterKey] = true;
                        console.log(`Got it! Vote cast for ${candidate}`);
                        break;
                    }
                }
            } else {
                console.log(" That's not a valid choice!");
            }
        }

        displayResults(round2Votes);
        
        const maxVotes = Math.max(...Object.values(round2Votes));
        const winners = Object.entries(round2Votes).filter(([_, v]) => v === maxVotes).map(([c]) => c);

        if (winners.length === 1) {
            console.log("\n We have a winner!");
            fancyPrint(`Congratulations ${winners[0]}!`);
            break;
        } else {
            console.log("\n Still tied! Let's go again...");
        }
    }
}

async function main() {
    console.log("\n" + "=".repeat(50));
    fancyPrint("Welcome to USA Voting System 2024");
    console.log("=".repeat(50) + "\n");
    
    const candidates = registerCandidates();
    const votes = await castVote(candidates);
    const finalVotes = displayResults(votes);

    const maxVotes = Math.max(...Object.values(finalVotes));
    const winners = Object.entries(finalVotes).filter(([_, v]) => v === maxVotes).map(([c]) => c);

    if (winners.length > 1) {
        console.log("\n Looks like we have a tie!");
        await runSecondRound(winners);
    } else {
        console.log("\n We have a winner!");
        fancyPrint(`Congratulations ${winners[0]}!`);
    }

    readline.close();
}

main();
